### DISCUSSION TITLE
New to GitHub, Uploading Files Won't Work

### DISCUSSION THREAD

**Discussion by antamongtitans** [community/community#64909] [New to GitHub]
Hello to the community!

I've finally reached the point in time in my career where people are looking for a GitHub of mine to see my work. I'm new to this, and I'm trying to upload. I'm running this on Mac, 2023 16-inch, running macOS Ventura 13.3.1(a). In the past, on an old laptop, I messed around with GitHub and Python and I've had to spend several years fixing the mess I made.

My issue is thus: I followed a video on YouTube of how to upload files to a repository, and everything worked up to 

git push -u origin main

It then gave me the error:

remote: Permission to antamongtitans/PanthersCodingProject.git denied to antamongtitans.
fatal: unable to access 'https://github.com/antamongtitans/PanthersCodingProject.git/': The requested URL returned error: 403

 I went looking for a solution. I set up both a fine-grain and a classic personal access token. I went through my git config --list and erased my old name and email. I deleted the Keychain Access entries for GitHub. I tried readding the proper information to a new Keychain Access item. I tried switching from private repository to public, and vice versa. Nothing has worked.

What am I doing wrong? How can I upload it? Please help!

**Comment by d-suter** [comment]
Hey antamongtitans,

I completely understand the frustration that comes with Git errors. The error you're seeing typically arises due to authentication or permission issues.

## **Step-by-Step Troubleshooting:**

1. **Authentication Method:**
    - GitHub has stopped supporting password authentication for push operations, hence, if you're using your password, you might want to switch to either SSH or using a Personal Access Token (PAT).
    
2. **Setting Up SSH (Optional if you prefer SSH over HTTPS):**
    - Check if you already have an SSH key:
        ```bash
        cat ~/.ssh/id_rsa.pub
        ```
      If this returns a key, you have one. If not, generate one:
        ```bash
        ssh-keygen -t rsa -b 4096 -C "your_email@example.com"
        ```
    - Start the ssh-agent in the background:
        ```bash
        eval "$(ssh-agent -s)"
        ```
    - Add your SSH private key to the ssh-agent:
        ```bash
        ssh-add ~/.ssh/id_rsa
        ```
    - Now, copy your public SSH key (the content you see when you `cat ~/.ssh/id_rsa.pub`) and go to GitHub > Settings > SSH and GPG keys > New SSH key, and paste it there.

3. **Using Personal Access Token (PAT) with HTTPS:**
    - Go to GitHub > Settings > Developer settings > Personal access tokens > Generate new token.
    - Make sure to select the required scopes. For pushing to a repository, the `repo` scope is essential.
    - After generation, copy the PAT. **It won’t be shown again!**
    - When you push and it asks for username and password, use your GitHub username and the PAT as the password.

4. **Change the Remote URL:**
    - If you are currently using HTTPS and wish to switch to SSH (or vice versa), you'll need to change the remote URL.
        ```bash
        git remote set-url origin git@github.com:antamongtitans/PanthersCodingProject.git
        ```
      (This is for SSH. For HTTPS, it will be `https://github.com/antamongtitans/PanthersCodingProject.git`)

5. **Retry the Push:**
    ```bash
    git push -u origin main
    ```
    At this point, if you're using SSH, it should directly push. If using HTTPS with a PAT, it will ask for credentials; provide your username and PAT as mentioned above.

6. **Clear Old Credentials:**
    - You mentioned about Keychain. Ensure there are no old GitHub credentials in there.
        - Open Keychain Access.
        - Search for `github.com`.
        - Delete any entries related to it.
    - If you're prompted for a username and password when you push, ensure you're using the latest credentials.

7. **Repository Ownership & Permissions:**
    - Ensure you have the proper permissions for the repository. If it’s a repo under an organization or someone else’s, you might need to be added as a collaborator or member.

8. **Try Using the Command Line Over GUI:**
    - If you're using any GUI tools, there might be hidden configurations affecting the operation. Always double-check using the terminal or command line.

9. **Cloning Again (as a last resort):**
    - Sometimes, the easiest way to solve these issues is to:
        1. Ensure all your changes are saved.
        2. Delete the local repository.
        3. Clone it fresh.
        4. Add your changes.
        5. Commit and push.

Hopefully, by following these steps, your issue will be resolved. If you're still having problems, don't hesitate to reach out again. We're here to help! Best of luck with your coding projects on GitHub.

**Reply by antamongtitans** [reply]
I'm working with HTTPS. The repository was created by me, as were all generated PATs. My Keychain has no github.com items.

I've checked my git config --list, and it is the following:

credential.helper=osxkeychain
init.defaultbranch=main
filter.media.clean=git-media-clean %f
filter.media.smudge=git-media-smudge %f
core.repositoryformatversion=0
core.filemode=true
core.bare=false
core.logallrefupdates=true
core.ignorecase=true
core.precomposeunicode=true
remote.origin.url=https://github.com/antamongtitans/PanthersCodingProject.git
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
user.name=antamongtitans
user.email=peterbrizardlarsen@gmail.com
user.credential.username=antamongtitans

I've generated 3 PATs so far, the last of which I made sure had repo selected. I created the repository.

The PAT with repo selected has given me the new error: 

remote: Support for password authentication was removed on August 13, 2021.
remote: Please see https://docs.github.com/en/get-started/getting-started-with-git/about-remote-repositories#cloning-with-https-urls for information on currently recommended modes of authentication.
fatal: Authentication failed for 'https://github.com/antamongtitans/PanthersCodingProject.git/'

No longer getting the 403 error, but it thinks to PAT is a password and is rejecting it. The PAT without repo selected is getting the following error:

remote: Write access to repository not granted.
fatal: unable to access 'https://github.com/antamongtitans/PanthersCodingProject.git/': The requested URL returned error: 403

I have all 3 PATs written down. The first is the new fine-grained option, gave the same errors as PAT 2, which is a classic but only has gists selected. The third is a classic with everything, including repo, selected.

What can I do now?
